snakemake/remote/GS.py

```diff
@@ -95,12 +126,27 @@ def size(self):
             return self._iofile.size_local
 
     def download(self):
-        if self.exists():
-            os.makedirs(os.path.dirname(self.local_file()), exist_ok=True)
-            self.blob.download_to_filename(self.local_file())
+        if not self.exists():
+            return None
+
+        # Create the directory for the intended file
+        os.makedirs(os.path.dirname(self.local_file()), exist_ok=True)
+
+        # Continue trying to download until hash matches
+        while not os.path.exists(self.local_file()):
```

Comment:
Instead of checking existence on disk, have a boolean variable that becomes true once the file is there or so. THis way, we save the additional existence checks, which can play a role when downloading many files in parallel.

Reply:
I almost implemented it this way! I changed at the last minute because I didn't like the is, else in the loop. I didn't think of this. Now fixed!